httpload/cmd/httpload/main.go

```diff
@@ -0,0 +1,440 @@
+	theRequest := c.url
+
+	if theRequest == "" {
+		return nil, nil, errors.Errorf("no URL provided")
+	}
+
+	out := make(chan string)
+	errc := make(chan error, 1) // unused
+
+	go func() {
+		defer close(out)
+		defer close(errc)
+
+		proceed := func(i int) bool {
+			// if the duration is defined we emit requests indefinitely
+			isDurationDefined := (c.appDuration > 0)
+			// if no duration is defined, emit -n requests
+			isLoopUnfinished := (i < c.nRequests)
+			isUnfinished := isDurationDefined || isLoopUnfinished
+			return isUnfinished
+		}
+
+		for i := 0; proceed(i); i++ {
+			select {
+			case out <- theRequest:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return out, errc, nil
+}
+
+// LoadTester represents an instance of the load testing logic
+type LoadTester struct {
+	results Results
+}
+
+func newLoadTesterFromConfig(config *Config) LoadTester {
+	loadTester := LoadTester{}
+	return loadTester
+}
+
+func (lt *LoadTester) writeResults(w io.Writer) {
+	fmt.Fprintf(w, "\n%s\n\n", "Summary:")
+	lt.dumpTimings(w)
+
+	fmt.Fprintf(w, "\n%s\n\n", "Status code distribution:")
+	lt.dumpStatuses(w)
+}
+
+func (lt *LoadTester) genLoadRequest(ctx context.Context, in <-chan string) (<-chan Result, <-chan error, error) {
+	out := make(chan Result)
+	errc := make(chan error, 1)
+
+	go func() {
+		defer close(out)
+		defer close(errc)
+
+		for url := range in {
+			req, err := http.NewRequest("GET", url, nil)
+			if err != nil {
+				errc <- err
+				return
+			}
+
+			req = req.WithContext(ctx)
+
+			start := time.Now()
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				errc <- err
+				return
+			}
+			defer resp.Body.Close()
+
+			elapsed := time.Since(start)
+			code := resp.StatusCode
+
+			select {
+			case out <- Result{code, elapsed}:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return out, errc, nil
+}
+
+func (lt *LoadTester) genLoadRequests(ctx context.Context, config *Config, in <-chan string) (<-chan Result, <-chan error, error) {
+	out := make(chan Result)
+	errc := make(chan error, 1)
+
+	var wg sync.WaitGroup
+	numWorkers := config.nWorkers
+	wg.Add(numWorkers)
+
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			resultc, errcr, err := lt.genLoadRequest(ctx, in)
+			if err != nil {
+				errc <- err
+				return
+			}
+
+			for result := range resultc {
+				select {
+				case out <- result:
+				case <-ctx.Done():
+					return
+				}
+			}
+			for err := range errcr {
+				select {
+				case errc <- err:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+
+		close(out)
+		close(errc)
+	}()
+
+	return out, errc, nil
+}
+
+func (lt *LoadTester) genResults(ctx context.Context, in <-chan Result) (<-chan error, error) {
+	errc := make(chan error, 1) // signals goroutine completion
+
+	go func() {
+		defer close(errc)
+
+		for result := range in {
+			lt.results = append(lt.results, result)
```

Comment:
race condition : you append to the results slice from multiple goroutines at once

Reply:
It is only 1 goroutine reading from the `in` channel. I added a `mutex`, but isn't it the channel purposes to avoid using synchronization primitives? Am i missing something?

Reply:
sorry my bad I didn't express myself correctly, the data race is not due to multiple concurrent writes but concurrent read/writes, specifically:

```
==================
WARNING: DATA RACE
Read at 0x00c00010dbc0 by main goroutine:
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).DumpStatuses()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:40 +0x108
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).WriteResults()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:109 +0x155
  main.trueMain()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/cmd/httpload/main.go:61 +0x4ba
  main.main()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/cmd/httpload/main.go:75 +0x17a

Previous write at 0x00c00010dbc0 by goroutine 15:
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).AppendResults.func1()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:208 +0x13a

Goroutine 15 (finished) created at:
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).AppendResults()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:204 +0x7d
  github.com/rgianassi/learning/go/httpload/pkg/loader.(*LoadTester).RunLoaderPipeline()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/pkg/loader/loadtester.go:294 +0x3b3
  main.trueMain.func1()
      /home/aurelien/godev/src/github.com/rgianassi/learning/httpload/cmd/httpload/main.go:50 +0x46
```

Reply:
`loadtester.go:40` and `loadtester.go:208` are called at the same time from 2 different goroutines, hence the data race

Reply:
Maybe, I am not understanding something obvious, but `loadtester.go:208` is called inside `AppendResults` which is a stage of the pipeline in `RunLoaderPipeline` which blocks until the pipeline is ended (because of `<-done` statement). On the other end, `loadtester.go:40` reads when the pipeline is done, so no worker is writing to it anymore (and it must not, otherwise we have partial results). How do you get the data race condition? I tried, but I am not able to.